WCSimFLOWER.cpp

```diff
@@ -0,0 +1,441 @@
+    if(abs(z) > fTopBottomDistanceLo && abs(z) < fTopBottomDistanceHi)
+      // TODO: deal with PMTs at the edge of the top/bottom
+      ratio = nearbyHits / 5.0;
+    else
+      ratio = nearbyHits / 8.0;
+
+    if (ratio == 0) {
+      occupancy = 1.0;
+    } else if (ratio < 1) {
+      occupancy = log(1 / (1-ratio)) / ratio; // from Poisson statistics
+    } else {
+      occupancy = 3.0;
+    }
+
+    // correct for delayed hits (e.g. due to scattering)
+    lateHits = (fNMaxLong - fNMaxShort - (fNWorkingPMTs * fDarkRate * fShortDuration)) / fNMaxShort;
+
+    // substract dark noise hits
+    darkNoise = (fNWorkingPMTs * fDarkRate * fShortDuration) / fNMaxShort;
+
+    // calculate effective coverage to correct for photoCoverage
+    // this depends on angle of incidence, see Fig. 4.5 (left) of http://www-sk.icrr.u-tokyo.ac.jp/sk/_pdf/articles/2016/doc_thesis_naknao.pdf
+    // TODO: take into account azimuthal dependence as show in Fig. 4.5 (right); currently assumes phi=0
+    float dotProduct = pmt.GetOrientation(0)*(fVertex[0] - x) +
+      pmt.GetOrientation(1)*(fVertex[1] - y) +
+      pmt.GetOrientation(2)*(fVertex[2] - z);
+    float theta = acos( dotProduct / fDistanceShort[i]) * 180 / 3.14159265;
+    if (theta > 89.99) theta = 0; // we have apparently mis-reconstructed the vertex, so let's set ...
+    if (theta < 0) theta = 0; // ... the coverage to the most likely value of 0.4 (i.e. theta < 40 degrees)
+    
+    photoCoverage = 1 / WCSimFLOWER::fEffCoverages[int(theta/10)];
+    if (fDetector == kHyperK20)
+      photoCoverage *= 38448/float(19462); // ratio of number of PMTs is not exactly 2
+
+    // correct for scattering in water
+    waterTransparency = exp(fDistanceShort[i] / fLambdaEff);
+
+    nEffHit = (occupancy + lateHits - darkNoise) * photoCoverage * waterTransparency;
+    fNEff += nEffHit;
+    // ad-hoc modification for low photocoverage
+    nEffHit2 = (pow(occupancy, 1.4) + lateHits - darkNoise) * photoCoverage * waterTransparency;
+    fNEff2 += nEffHit2;
+    
+    if (fVerbose > 1) {
+      std::cout << "\n*** PMT hit #" << i << " on tube " << tubeID << " ***************************************\n";
+      std::cout << "occupancy (ratio of hits in 3x3 grid): " << occupancy << " (" << ratio << ")\n";
+      std::cout << "lateHits:  " << lateHits << "\n";
+      std::cout << "darkNoise: " << darkNoise << "\n";
+      std::cout << "photoCoverage: " << photoCoverage << "\n";
+      std::cout << "waterTransparency: " << waterTransparency << "\n";
+      std::cout << "nEff for this 1 hit: " << nEffHit << "\n";
+    }
+  } // i //end of loop over hits in 50 ns interval
+
+
+ // correct for dead PMTs; convert nWorkingPMTs to float because integer division is inaccurate
+  fNEff  *= fNPMTs / float(fNWorkingPMTs);
+  fNEff2 *= fNPMTs / float(fNWorkingPMTs);
+
+  if(fVerbose) {
+    std::cout << endl << "***************************************" << endl
+	      << "Average nEff for all hits: " << fNEff << endl
```

Comment:
```suggestion
	      << "nEff for this event: " << fNEff << endl
```

It’s the sum over all hits in the event, not an average.